adagrams/game.py

```diff
@@ -1,11 +1,174 @@
+import random
+
+LETTER_POOL_W_VALUES = {
+    'A': {
+        'qty': 9, 
+        'value': 1
+    },
+    'B': {
+        'qty': 2,
+        'value': 3
+    },
+    'C': {
+        'qty': 2, 
+        'value': 3
+    },
+    'D': {
+        'qty': 4, 
+        'value': 2
+    },
+    'E': {
+        'qty': 12,
+        'value': 1
+    },
+    'F': {
+        'qty': 2, 
+        'value': 4
+    },
+    'G': {
+        'qty': 3, 
+        'value': 2
+    },
+    'H': {
+        'qty': 2, 
+        'value': 4
+    },
+    'I': {
+        'qty': 9,
+        'value' :1
+    },
+    'J': {
+        'qty': 1, 
+        'value': 8
+    },
+    'K': {
+        'qty': 1, 
+        'value': 5
+    },
+    'L': {
+        'qty': 4, 
+        'value': 1
+    },
+    'M': {
+        'qty': 2, 
+        'value': 3
+    },
+    'N': {
+        'qty': 6, 
+        'value': 1
+    },
+    'O': {
+        'qty': 8,
+        'value': 1
+    },
+    'P': {
+        'qty': 2, 
+        'value': 3
+    },
+    'Q': {
+        'qty': 1, 
+        'value': 10
+    },
+    'R': {
+        'qty': 6,
+        'value': 1
+    },
+    'S': {
+        'qty': 4, 
+        'value': 1
+    },
+    'T': {
+        'qty': 6, 
+        'value': 1
+    },
+    'U': {
+        'qty': 4, 
+        'value': 1
+    },
+    'V': {
+        'qty': 2, 
+        'value': 4
+    },
+    'W': {
+        'qty': 2, 
+        'value': 4
+    },
+    'X': {
+        'qty': 1, 
+        'value': 8
+    },
+    'Y': {
+        'qty': 2, 
+        'value': 4
+    },
+    'Z': {
+        'qty': 1,
+        'value': 10
+    }
+}
+
 def draw_letters():
-    pass
+    letters = []
+    frequency_of_letters = []
+    for letter, frequency in LETTER_POOL_W_VALUES.items():
+        letters.append(letter)
+        frequency_of_letters.append(frequency['qty'])
+    # random.sample returns a new list containing elements from the population
+    # while leaving the original population unchanged. counts controls the
+    # probability of an element within the sample. in this case, the greater
+    # the quantity of tiles for a letter, the higher probability that letter
+    # will be part of the random sample.
+    random_letters_list = random.sample(letters, counts =\
+        frequency_of_letters, k = 10)
+    return random_letters_list
 
 def uses_available_letters(word, letter_bank):
-    pass
+    copy_letter_bank = letter_bank.copy()
+    for letter in word.upper():
```

Comment:
This works perfectly well! One thing to note is that you're looping through every letter in word on line 127 and then you're using the `in` operator to check if the letter is in the `copy_letter_bank` on line 128. Recall that using the in operator has O(n) time complexity, as does your for loop, so we've got O(n^2).

Rather than using a list of the letters in the hand, could we build a helper data structure (like a dictionary) that could let us look up the number of tiles remaining of each type?

If you have a frequency dict and then check if a key is in the dict, then you can leverage the fact that dicts constant time lookup (in contrast to the list's linear time look up).